### PR TITLE
Migrate to laminas/laminas-coding-standard v2

### LINKED ISSUE #5: Update Coding Standard to v2
There are a lot of new conventions that are not covered by the current phpcs ruleset of [zend-coding-standard](/zendframework/zend-coding-standard) (v1.x).

Some examples are (incomplete list):

- Whitespace after cast operators
- Whitespace after not operator (`!`)
- Whitespaces before and after ternary operators
- Trailing comma for the last element in array declarations

A new coding style and tooling is currently work in progress at zendframework/zend-coding-standard#5. When this PR is merged and v2 is released this zend-di should integrate it.

Depending on my availability, I'll perform the tests as suggested in the mentioned PR.

---

Originally posted by @tux-rampage at https://github.com/zendframework/zend-di/issues/39

### PR DESCRIPTION
**Pull request by gennadigennadigennadi**
|    Q          |   A
|-------------- | ------
| Bugfix        | yes
| QA            | yes

### Description

Started working on migration Code style to v2.

### CODE REVIEW

File: composer.json
Comment (tux-rampage) [inline composer.json]: This is already targeted in PR #15 
Could you explain. why you changed it here? To me it seems that it is not needed to update the coding style.

File: phpstan.neon
Comment (tux-rampage) [inline phpstan.neon]: The same as for `composer.json`: This is already targeted in PR #15
Is it required for applying the CS?

File: src/CodeGenerator/InjectorGenerator.php
Comment (tux-rampage) [inline src/CodeGenerator/InjectorGenerator.php:77]: This looks a bit off @geerteltink and @michalbundyra - Is this how it should look like for multiline doc-tags?

File: composer.json
Comment (tux-rampage) [inline composer.json]: This seems to be already addressed in PR #15 
Is this (Version bump for phpstan) needed for applying the CS or was it committed by accident? To me it seems like this is not necessary here.

File: phpstan.neon
Comment (tux-rampage) [inline phpstan.neon]: Same as for `composer.json`. Is this required. It seems like a side effect of the phpstam version bump.
If not required, please revert

File: src/CodeGenerator/InjectorGenerator.php
Comment (tux-rampage) [inline src/CodeGenerator/InjectorGenerator.php:77]: These comments look a bit off. Was this applied by the coding standard's auto-fixer?

File: src/Config.php
Comment (tux-rampage) [inline src/Config.php:184]: This introduces a BC break: When some Stringable (i.e. a class with a `__toString()` method) is placed in the config it will now change behavior. Prior this change it would be accepted and stringified. After this it becomes `null`.

File: src/Container/AutowireFactory.php
Comment (tux-rampage) [inline src/Container/AutowireFactory.php]: This return type is incorrect, it shoud be `object`

File: src/Container/AutowireFactory.php
Comment (tux-rampage) [inline src/Container/AutowireFactory.php]: Same as above: Return type is `object`

File: src/Resolver/AbstractInjection.php
Comment (tux-rampage) [inline src/Resolver/AbstractInjection.php:2]: Why should this file be ignored from CS?

File: src/Resolver/DependencyResolverInterface.php
Comment (tux-rampage) [inline src/Resolver/DependencyResolverInterface.php:37]: This looks also a bit off. Was this also applied by the auto-fixer?

File: test/InjectorTest.php
Comment (tux-rampage) [inline test/InjectorTest.php:336]: This looks lie auto fixer. @geerteltink and @michalbundyra  This looks really wired and is hard to read for multi dimensional arrays. Is this the intended effect/style?

File: test/TestAsset/Constructor/OptionalArguments.php
Comment (tux-rampage) [inline test/TestAsset/Constructor/OptionalArguments.php]: This changes the constructor signature of this test asset. Consider using doc-block tags instead.

**Review by tux-rampage — CHANGES_REQUESTED** [review]
Thank you very much for taking care of this. :+1:
There are some pieces that might need adjustments. Could you please check my feedback on the diff?

**Comment by tux-rampage** [comment]
There are duplicate remarks, github was not showing them anymore after I temporarly navigated away from the review :facepalm: sorry about that.

The builds are failing for PHP `7.1` which is IMHO not a big deal - it's EOL anyways.

I'd suggest to wait for PR #15 (which is almost ready) to be merged since this PR must be rebased anyways afterward.

File: phpstan.neon
Comment (tux-rampage) [reply phpstan.neon]: Duplicate - I'll mark as resolved, sorry about this.

File: composer.json
Comment (tux-rampage) [reply composer.json]: Duplicate - I'll mark as resolved, sorry about this.

File: src/CodeGenerator/InjectorGenerator.php
Comment (tux-rampage) [reply src/CodeGenerator/InjectorGenerator.php:77]: Duplicate - I'll mark as resolved, sorry about this.

File: src/Resolver/DependencyResolverInterface.php
Comment (gennadigennadigennadi) [reply src/Resolver/DependencyResolverInterface.php:37]: Those multiline DocBlock got "fixed" automatically.

File: src/Resolver/AbstractInjection.php
Comment (gennadigennadigennadi) [reply src/Resolver/AbstractInjection.php:2]: Phpcs fails because of the deprecation warning.

**Review by boesing — CHANGES_REQUESTED** [review]
I think, the following TODOs are open:
- remove phpstan changes
- do not add native php typehints to `protected` or `public` methods (unless they're marked as `final`)  for classes which are not final as this will require a major version bump which should not be needed to just apply a new coding standard

Besides that, well done.

File: src/CodeGenerator/AbstractInjector.php
Comment (boesing) [inline src/CodeGenerator/AbstractInjector.php:37]: Please revert his change as it introduces BC break.

File: src/CodeGenerator/InjectorGenerator.php
Comment (boesing) [inline src/CodeGenerator/InjectorGenerator.php:82]: Please revert his change as it introduces BC break.

File: src/Container/AutowireFactory.php
Comment (boesing) [inline src/Container/AutowireFactory.php]: Please revert his change as it introduces BC break.

File: src/Container/AutowireFactory.php
Comment (boesing) [inline src/Container/AutowireFactory.php]: Please revert his change as it introduces BC break.

File: src/Container/ServiceManager/AutowireFactory.php
Comment (boesing) [inline src/Container/ServiceManager/AutowireFactory.php:27]: Please revert his change as it introduces BC break.

File: src/Container/ServiceManager/AutowireFactory.php
Comment (boesing) [inline src/Container/ServiceManager/AutowireFactory.php:49]: Please revert his change as it introduces BC break.

File: src/Definition/RuntimeDefinition.php
Comment (boesing) [inline src/Definition/RuntimeDefinition.php:34]: Please revert his change as it introduces BC break.

File: src/Exception/ClassNotFoundException.php
Comment (boesing) [inline src/Exception/ClassNotFoundException.php:16]: Please revert his change as it introduces BC break.

File: src/Injector.php
Comment (boesing) [inline src/Injector.php:64]: Please revert his change as it introduces BC break.

File: src/Injector.php
Comment (boesing) [inline src/Injector.php:65]: Please revert his change as it introduces BC break.

File: src/Injector.php
Comment (boesing) [inline src/Injector.php:66]: Please revert his change as it introduces BC break.

File: src/Injector.php
Comment (boesing) [inline src/Injector.php:67]: Please revert his change as it introduces BC break.

File: src/Resolver/AbstractInjection.php
Comment (boesing) [reply src/Resolver/AbstractInjection.php:2]: Maybe we should just disable the rule then instead of disabling the whole file from being checked.

File: src/Resolver/DependencyResolver.php
Comment (boesing) [inline src/Resolver/DependencyResolver.php]: We should only apply codestyle changes in this PR. 
If this would be an improvement, please create another PR with this change.

Please revert this change in this PR.

File: src/Resolver/ValueInjection.php
Comment (boesing) [inline src/Resolver/ValueInjection.php:126]: Was that applied by the codestyle? Imho, we should stick with `__CLASS__` constant.

**Review by boesing — CHANGES_REQUESTED** [review]

File: src/Resolver/DependencyResolverInterface.php
Comment (tux-rampage) [reply src/Resolver/DependencyResolverInterface.php:37]: I see, thanks.

@boesing should it look like this. Maybe this is some edge case the auto-fixer ran into? Shouldn't this be just 4 spaces of indentation for each new line?

File: src/CodeGenerator/AbstractInjector.php
Comment (gennadigennadigennadi) [reply src/CodeGenerator/AbstractInjector.php:37]: These additional ? were added by the phpcbf, but I added the others Typehints. As soon as #15 is merged I will revert some of the Typehints that are breaking changes.

Comment (tux-rampage) [reply src/CodeGenerator/AbstractInjector.php:37]: Are you sure that this breaks BC? This is something the coding standard dictates. Afaik these types are already nullable   since they declare `null` as default value.

Edit: @boesing

File: src/Resolver/DependencyResolver.php
Comment (gennadigennadigennadi) [reply src/Resolver/DependencyResolver.php]: You are right. 
At first the phpcbf changes in the X.php/Y.php  the a Typehint to self and the ```$paramInfo->getType()``` returned the string `self` instead of the fqcn. Therefore I added the ```// phpcs:ignoreFile``` to X.php/Y.php I I forgot to revert these changes.

File: src/Resolver/ValueInjection.php
Comment (gennadigennadigennadi) [reply src/Resolver/ValueInjection.php:126]: phpcbf did this.

File: src/Resolver/DependencyResolverInterface.php
Comment (boesing) [reply src/Resolver/DependencyResolverInterface.php:37]: > @boesing should it look like this. Maybe this is some edge case the auto-fixer ran into? Shouldn't this be just 4 spaces of indentation for each new line?

TBH - I dont know. I'd say, if the phpcbf fixes the code like this and it wasnt applied manually, we should keep this as is and either wait until there is a bugfix upstream. 🤷‍♂️ 
Feel free to report that.

File: test/InjectorTest.php
Comment (boesing) [reply test/InjectorTest.php:336]: As explained [here](https://github.com/laminas/laminas-di/pull/16#discussion_r499811386), we should keep this as is as it will probably re-formatted again by the beautifier.

File: src/CodeGenerator/AbstractInjector.php
Comment (boesing) [reply src/CodeGenerator/AbstractInjector.php:37]: @gennadigennadigennadi right, in `__construct` thats not an issue (as it is treated different to other methods).

@tux-rampage nope, not sure. Actually, `__construct` signature changes do not have implications as tested [here](https://3v4l.org/bW2O8)

Methods other than `__construct` will lead to [this](https://3v4l.org/kmB34) error in classes which use inheritance

Comment (tux-rampage) [reply src/CodeGenerator/AbstractInjector.php:37]: @boesing your example already leads to an error with the current codebase because it declares `null` as param default: https://3v4l.org/6GSPb

So `foo(type $var = null)` is an implicit `foo(?type = null)`. The coding style just enforces to make this obvious. So as far as I can tell this does not break BC or am I missing something else?

File: src/Resolver/DependencyResolverInterface.php
Comment (tux-rampage) [reply src/Resolver/DependencyResolverInterface.php:37]: Agree

**Comment by tux-rampage** [comment]
@gennadigennadigennadi thanks for your patience. I've merged #15, could you please rebase your branch on `3.2.x` and address the feedback? :v:

**Comment by gennadigennadigennadi** [comment]
I rebased yesterday and I am going to work the next few days on the feedback.

File: src/CodeGenerator/InjectorGenerator.php
Comment (gennadigennadigennadi) [reply src/CodeGenerator/InjectorGenerator.php:82]: those ? are added by cs-fix

File: src/Container/ServiceManager/AutowireFactory.php
Comment (gennadigennadigennadi) [reply src/Container/ServiceManager/AutowireFactory.php:27]: those ? are added by cs-fix

File: src/Injector.php
Comment (gennadigennadigennadi) [reply src/Injector.php:67]: those ? are added by cs-fix

File: src/Container/AutowireFactory.php
Comment (gennadigennadigennadi) [reply src/Container/AutowireFactory.php]: I removed the string Typehint. The ? before array was added by cs-fix.

File: src/Injector.php
Comment (gennadigennadigennadi) [reply src/Injector.php:66]: those ? are added by cs-fix

File: src/CodeGenerator/InjectorGenerator.php
Comment (gennadigennadigennadi) [reply src/CodeGenerator/InjectorGenerator.php:77]: yup. At least I would not Style code like this ;-)

File: src/Config.php
Comment (gennadigennadigennadi) [reply src/Config.php:184]: This one was also made by php-cs. check if $preference is a class and has __toString then call __toString by ourselves?

File: src/Container/AutowireFactory.php
Comment (tux-rampage) [inline src/Container/AutowireFactory.php]: Even though this would not break BC, please don't remove the `string` typehint, that would remove type safety.

File: src/LegacyConfig.php
Comment (tux-rampage) [inline src/LegacyConfig.php]: This should be an array of `mixed`

```suggestion
     * @return mixed[]
```

File: src/LegacyConfig.php
Comment (tux-rampage) [inline src/LegacyConfig.php]: This is a BC break. Please remove this type hint.

Edit: Event though this is private, we're not on psalm with full static analysis, yet and for phpstan there is a baseline in place. So it may break something undetected.

File: test/Resolver/DependencyResolverTest.php
Comment (tux-rampage) [inline test/Resolver/DependencyResolverTest.php]: Why do you remove this parameter?

**Review by tux-rampage — CHANGES_REQUESTED** [review]
@gennadigennadigennadi Thank you very much for rebasing and adressing the feedback. There are still a view small things that would introduce breaking changes. Could you please check these?
Otherwise great work :+1:

File: test/Resolver/DependencyResolverTest.php
Comment (gennadigennadigennadi) [reply test/Resolver/DependencyResolverTest.php]: https://github.com/laminas/laminas-di/blob/3.2.x/test/Resolver/DependencyResolverTest.php#L78
Ist not used inside mockClassDefinition. 
But as you said only code style changes, no refactorings. ;-). Im going to revert those last changes and I should be good.

File: src/Container/AutowireFactory.php
Comment (gennadigennadigennadi) [reply src/Container/AutowireFactory.php]: this one is a mistake, I thought this one was added by my commits

**Review by tux-rampage — APPROVED** [review]
LGTM. Great work, thank you very much. :+1: 

@boesing Can you please check if your feedback is was addressed or if there are any objections on your side?

File: src/Container/AutowireFactory.php
Comment (tux-rampage) [reply src/Container/AutowireFactory.php]: No worries, those slip through easily on such a large changeset. That's why we do reviews. :wink:

File: src/CodeGenerator/AbstractInjector.php
Comment (boesing) [reply src/CodeGenerator/AbstractInjector.php:37]: I've re-checked this again and you are right... adding the nullable-flag to a typehint is not a BC break. I probably miss-understood something... Damn!

@gennadigennadigennadi sorry for bothering you with this.

**Review by boesing — APPROVED** [review]

**Comment by boesing** [comment]
LGTM aswell. I definitely was wrong with that nullable flag. I was totally convinced it introduces BC break but it doesn't. Sorry for bothering you two...

File: src/Config.php
Comment (gennadigennadigennadi) [reply src/Config.php:184]: @boesing and @tux-rampage this problem still has to be fixed!
What do you guys think how to solve it?

Comment (geerteltink) [reply src/Config.php:184]: @gennadigennadigennadi You can run `vendor/bin/phpcs -s` and check which rule is trying to change it. When you know, you can add `// phpcs:ignore some.phpcs.rule` in front of it so it won't get changed.

**Pull request merged**
